src/algorithms/cf.py

```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
+    """
+    Determine the predicted rating of every user-item combination using Naive Bayes MAP.
+    Pai     : predicted rating for user a on item i
+    P(r|a,i): Naive Bayes that computes the probability of rating r for a given user a on item i
+    Pai = Argmax(r=1 to 5) P(r|a,i)
+
+    Assumption:
+    Since Naive Bayes can be defined as P(r|u,i) = ( (P(r|u) * P(r|i)) / P(r) ) * ( (P(u) * P(i)) / P(u, i)), 
+    we make the assumption that the latter part of the multiplication, ( (P(u) * P(i)) / P(u, i)), can be ignored.
+
+    Args:
+        training_data: the data used to train the RecSys algorithm in the format of a RDD of [ (userId, itemId, actualRating) ]
+
+    Returns:
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    def calculate_bayes_map(value):
+        # extract the bayes_prob 
+        bayesProbList = [x[1] for x in value]
+        
+        # define the argmax, return the index
+        argmax = bayesProbList.index(max(bayesProbList))
+        
+        return argmax
+
+    if computeFromScratch:
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
+            raise Exception("ERROR: ui_allBayesProb is not defined although user specified not to calculate ui_allBayesProb not from scratch.")
+
+    return ui_allBayesProb.mapValues(calculate_bayes_map)
+
+
+def calc_naive_bayes_mse(training_data, sc):
+    """
+    Determine the predicted rating of every user-item combination using Naive Bayes MSE.
+    Pai     : predicted rating for user a on item i
+    P(r|a,i): Naive Bayes that computes the probability of rating r for a given user a on item i
+    Pai = Sum of (r * P(r|a,i)) from r=1 to 5
+
+    Assumption:
+    Since Naive Bayes can be defined as P(r|u,i) = ( (P(r|u) * P(r|i)) / P(r) ) * ( (P(u) * P(i)) / P(u, i)), 
+    we make the assumption that the latter part of the multiplication, ( (P(u) * P(i)) / P(u, i)), can be ignored.
+
+    Args:
+        training_data: the data used to train the RecSys algorithm in the format of a RDD of [ (userId, itemId, actualRating) ]
+
+    Returns:
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    def calculate_bayes_mse(value):
+        predicted = 0.
+        for rating, bayes_prob in value:
+            predicted += rating * bayes_prob
+        return predicted
+
+    if computeFromScratch:
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
+            raise Exception("ERROR: ui_allBayesProb is not defined although user specified not to calculate ui_allBayesProb not from scratch.")
+
+    return ui_allBayesProb.mapValues(calculate_bayes_mse)
+
+
+def calc_naive_bayes_mae(training_data, sc):
```

Comment:
Noted that it requires additional arguments: (training_data, sc, computeFromScratch=True, ui_allBayesProb=None)